src/core/ext/filters/http/server/http_server_filter.cc

```diff
@@ -28,306 +28,126 @@
-  grpc_closure recv_trailing_metadata_ready;
-  grpc_closure* original_recv_trailing_metadata_ready;
-  grpc_error_handle recv_trailing_metadata_ready_error;
-  bool seen_recv_trailing_metadata_ready = false;
-};
-
-struct channel_data {
-  bool surface_user_agent;
-  bool allow_put_requests;
-};
-
-}  // namespace
-
-static grpc_error_handle hs_filter_outgoing_metadata(grpc_metadata_batch* b) {
-  if (grpc_core::Slice* grpc_message =
-          b->get_pointer(grpc_core::GrpcMessageMetadata())) {
-    *grpc_message = grpc_core::PercentEncodeSlice(
-        std::move(*grpc_message), grpc_core::PercentEncodingType::Compatible);
-  }
-  return GRPC_ERROR_NONE;
-}
+const grpc_channel_filter HttpServerFilter::kFilter =
+    MakePromiseBasedFilter<HttpServerFilter, FilterEndpoint::kServer,
+                           kFilterExaminesServerInitialMetadata>("http-server");
 
-static void hs_add_error(const char* error_name, grpc_error_handle* cumulative,
-                         grpc_error_handle new_err) {
-  if (new_err == GRPC_ERROR_NONE) return;
-  if (*cumulative == GRPC_ERROR_NONE) {
-    *cumulative = GRPC_ERROR_CREATE_FROM_COPIED_STRING(error_name);
+namespace {
+void FilterOutgoingMetadata(ServerMetadata* md) {
+  if (Slice* grpc_message = md->get_pointer(GrpcMessageMetadata())) {
+    *grpc_message = PercentEncodeSlice(std::move(*grpc_message),
+                                       PercentEncodingType::Compatible);
   }
-  *cumulative = grpc_error_add_child(*cumulative, new_err);
 }
+}  // namespace
 
-static grpc_error_handle hs_filter_incoming_metadata(grpc_call_element* elem,
-                                                     grpc_metadata_batch* b) {
-  grpc_error_handle error = GRPC_ERROR_NONE;
-  static const char* error_name = "Failed processing incoming headers";
+ArenaPromise<ServerMetadataHandle> HttpServerFilter::MakeCallPromise(
+    CallArgs call_args, NextPromiseFactory next_promise_factory) {
+  const auto& md = call_args.client_initial_metadata;
 
-  auto method = b->get(grpc_core::HttpMethodMetadata());
+  auto method = md->get(HttpMethodMetadata());
   if (method.has_value()) {
     switch (*method) {
-      case grpc_core::HttpMethodMetadata::kPost:
+      case HttpMethodMetadata::kPost:
         break;
-      case grpc_core::HttpMethodMetadata::kPut:
-        if (static_cast<channel_data*>(elem->channel_data)
-                ->allow_put_requests) {
+      case HttpMethodMetadata::kPut:
+        if (allow_put_requests_) {
           break;
         }
         ABSL_FALLTHROUGH_INTENDED;
-      case grpc_core::HttpMethodMetadata::kInvalid:
-      case grpc_core::HttpMethodMetadata::kGet:
-        hs_add_error(error_name, &error,
-                     GRPC_ERROR_CREATE_FROM_STATIC_STRING("Bad method header"));
-        break;
+      case HttpMethodMetadata::kInvalid:
+      case HttpMethodMetadata::kGet:
+        return Immediate(
+            ServerMetadataHandle(absl::UnknownError("Bad method header")));
```

Comment:
Should these failures actually be UNKNOWN?  I'm thinking INTERNAL would probably make more sense here.

(Looks like the existing code was already using UNKNOWN, so probably better to change this in a separate PR.  But this is a good indication of how much clearer the code is in the promise architecture: the fact that we were using the wrong status code here was very non-obvious in the old code.)